File: src/C3.Infrastructure/Preferences/LegacySettingsProfileReader.vb
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports System.Xml
Imports C3.Infrastructure.Updates

Namespace Preferences

    ''' <summary>
    ''' Reads only the scalar settings C3 has explicitly promised to migrate.
    ''' The source stream is opened read-only and is never changed or deleted.
    ''' </summary>
    Public NotInheritable Class LegacySettingsProfileReader

        Public Const MaximumProfileBytes As Long = 256L * 1024L

        Private Const SettingsSectionXPath As String =
            "/configuration/userSettings/Compact_Cassette_Catalogue.My.MySettings"

        Public Function Read(
                candidate As LegacySettingsProfileCandidate) As LegacySettingsProfileReadResult

            If candidate Is Nothing Then
                Throw New ArgumentNullException("candidate")
            End If

            Try
                Dim document As New XmlDocument()
                document.XmlResolver = Nothing

                Using stream As New FileStream(
                        candidate.FilePath,
                        FileMode.Open,
                        FileAccess.Read,
                        FileShare.Read)

                    If stream.Length > MaximumProfileBytes Then
                        Return LegacySettingsProfileReadResult.Failed(
                            candidate,
                            LegacySettingsProfileReadFailure.TooLarge,
                            "The legacy settings profile exceeds the 256 KiB safety limit.")
                    End If

                    Using reader As XmlReader = XmlReader.Create(stream, CreateSecureReaderSettings())
                        document.Load(reader)
                    End Using
                End Using

                Return ParseDocument(candidate, document)
            Catch ex As XmlException
                Return LegacySettingsProfileReadResult.Failed(
                    candidate,
                    LegacySettingsProfileReadFailure.MalformedXml,
                    "The legacy settings profile is not safe, well-formed XML.",
                    ex)
            Catch ex As IOException
                Return LegacySettingsProfileReadResult.Failed(
                    candidate,
                    LegacySettingsProfileReadFailure.Unavailable,
                    "The legacy settings profile could not be read.",
                    ex)
            Catch ex As UnauthorizedAccessException
                Return LegacySettingsProfileReadResult.Failed(
                    candidate,
                    LegacySettingsProfileReadFailure.Unavailable,
                    "Access to the legacy settings profile was denied.",
                    ex)
            Catch ex As System.Security.SecurityException
                Return LegacySettingsProfileReadResult.Failed(
                    candidate,
                    LegacySettingsProfileReadFailure.Unavailable,
                    "The legacy settings profile was blocked by the security policy.",
                    ex)
            End Try
        End Function

        Private Shared Function CreateSecureReaderSettings() As XmlReaderSettings
            Dim settings As New XmlReaderSettings()
            settings.DtdProcessing = DtdProcessing.Prohibit
            settings.XmlResolver = Nothing
            settings.MaxCharactersInDocument = MaximumProfileBytes
            settings.MaxCharactersFromEntities = 0L
            settings.IgnoreComments = True
            settings.IgnoreProcessingInstructions = True
            Return settings
        End Function

        Private Shared Function ParseDocument(
                candidate As LegacySettingsProfileCandidate,
                document As XmlDocument) As LegacySettingsProfileReadResult

            If document.DocumentElement Is Nothing OrElse
                    document.DocumentElement.Name <> "configuration" OrElse
                    document.DocumentElement.NamespaceURI.Length <> 0 Then
                Return InvalidStructure(candidate, "The configuration root is not the legacy C3 root.")
            End If

            Dim userSettingsNodes As XmlNodeList = document.SelectNodes("/configuration/userSettings")
            Dim sectionNodes As XmlNodeList = document.SelectNodes(SettingsSectionXPath)
            If userSettingsNodes Is Nothing OrElse userSettingsNodes.Count <> 1 OrElse
                    sectionNodes Is Nothing OrElse sectionNodes.Count <> 1 Then
                Return InvalidStructure(
                    candidate,
                    "The exact C3 1.x user settings section was not found once.")
            End If

            Dim section As XmlNode = sectionNodes.Item(0)
            Dim observedNames As New HashSet(Of String)(StringComparer.Ordinal)
            Dim recognizedCount As Integer

            Dim hasShowMessages As Boolean
            Dim showMessages As Boolean
            Dim hasDefaultDirectory As Boolean
            Dim defaultDirectory As String = Nothing
            Dim hasUpdatePolicy As Boolean
            Dim updatePolicy As UpdateCheckPolicy = UpdateCheckPolicy.Never
            Dim hasLastUpdateCheck As Boolean
            Dim lastUpdateCheck As DateTime = DateTime.MinValue

            For Each child As XmlNode In section.ChildNodes
                If child.NodeType <> XmlNodeType.Element OrElse child.Name <> "setting" Then
                    Continue For
                End If

                Dim setting As XmlElement = DirectCast(child, XmlElement)
                Dim settingName As String = setting.GetAttribute("name")
                If Not IsKnownSetting(settingName) Then
                    Continue For
                End If

                If Not observedNames.Add(settingName) Then
                    Return LegacySettingsProfileReadResult.Failed(
                        candidate,
                        LegacySettingsProfileReadFailure.DuplicateSetting,
                        "The legacy profile contains a duplicate supported setting: " & settingName)
                End If

                If setting.GetAttribute("serializeAs") <> "String" Then
                    Return InvalidValue(
                        candidate,
                        "The legacy setting has an unsupported serialization mode: " & settingName)
                End If

                Dim values As XmlNodeList = setting.SelectNodes("value")
                If values Is Nothing OrElse values.Count <> 1 OrElse HasElementChild(values.Item(0)) Then
                    Return InvalidValue(
                        candidate,
                        "The legacy setting does not contain one scalar value: " & settingName)
                End If

                Dim rawValue As String = values.Item(0).InnerText
                Select Case settingName
                    Case "showMessages"
                        If Not Boolean.TryParse(rawValue.Trim(), showMessages) Then
                            Return InvalidValue(candidate, "showMessages is not a Boolean value.")
                        End If
                        hasShowMessages = True
                    Case "defaultDirectory"
                        If rawValue.Length >
                                UserPreferencesSnapshot.MaximumDefaultDirectoryCharacters Then
                            Return InvalidValue(
                                candidate,
                                "defaultDirectory exceeds the supported safety limit.")
                        End If
                        defaultDirectory = rawValue
                        hasDefaultDirectory = True
                    Case "checkUpdates"
                        If Not TryParseUpdatePolicy(rawValue, updatePolicy) Then
                            Return InvalidValue(candidate, "checkUpdates is not a supported policy value.")
                        End If
                        hasUpdatePolicy = True
                    Case "lastUpdateCheck"
                        If Not TryParseLastUpdateCheck(rawValue, lastUpdateCheck) Then
                            Return InvalidValue(candidate, "lastUpdateCheck is not a supported date value.")
                        End If
                        hasLastUpdateCheck = True
                End Select

                recognizedCount += 1
            Next

            If recognizedCount = 0 Then
                Return InvalidStructure(candidate, "The legacy section contains no supported C3 settings.")
            End If

            Return LegacySettingsProfileReadResult.Succeeded(
                New LegacyUserSettingsProfile(
                    candidate,
                    hasShowMessages,
                    showMessages,
                    hasDefaultDirectory,
                    defaultDirectory,
                    hasUpdatePolicy,
                    updatePolicy,
                    hasLastUpdateCheck,
                    lastUpdateCheck))
        End Function

        Private Shared Function IsKnownSetting(name As String) As Boolean
            Return name = "showMessages" OrElse
                name = "defaultDirectory" OrElse
                name = "checkUpdates" OrElse
                name = "lastUpdateCheck"
        End Function

        Private Shared Function HasElementChild(node As XmlNode) As Boolean
            For Each child As XmlNode In node.ChildNodes
                If child.NodeType = XmlNodeType.Element Then
                    Return True
                End If
            Next
            Return False
        End Function

        Private Shared Function TryParseUpdatePolicy(
                rawValue As String,
                ByRef updatePolicy As UpdateCheckPolicy) As Boolean

            Dim legacyBoolean As Boolean
            If Boolean.TryParse(If(rawValue, String.Empty).Trim(), legacyBoolean) Then
                updatePolicy = If(
                    legacyBoolean,
                    UpdateCheckPolicy.Startup,
                    UpdateCheckPolicy.Never)
                Return True
            End If

            Return UpdateCheckSchedule.TryParseStored(rawValue, updatePolicy)
        End Function

        Private Shared Function TryParseLastUpdateCheck(
                rawValue As String,
                ByRef lastUpdateCheck As DateTime) As Boolean

            Dim normalized As String = If(rawValue, String.Empty).Trim()
            If normalized.Length = 0 Then
                lastUpdateCheck = DateTime.MinValue
                Return True
            End If

            Return DateTime.TryParse(
                normalized,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AllowWhiteSpaces Or DateTimeStyles.RoundtripKind,
                lastUpdateCheck)
        End Function

        Private Shared Function InvalidStructure(
                candidate As LegacySettingsProfileCandidate,
                message As String) As LegacySettingsProfileReadResult

            Return LegacySettingsProfileReadResult.Failed(
                candidate,
                LegacySettingsProfileReadFailure.InvalidStructure,
                message)
        End Function

        Private Shared Function InvalidValue(
                candidate As LegacySettingsProfileCandidate,
                message As String) As LegacySettingsProfileReadResult

            Return LegacySettingsProfileReadResult.Failed(
                candidate,
                LegacySettingsProfileReadFailure.InvalidValue,
                message)
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Preferences/LegacySettingsProfileReader.cs
using C3.Infrastructure.Updates;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Security;
using System.Xml;

namespace C3.Infrastructure.Preferences
{
	public sealed class LegacySettingsProfileReader
	{
		public const long MaximumProfileBytes = 262144L;

		private const string SettingsSectionXPath = "/configuration/userSettings/Compact_Cassette_Catalogue.My.MySettings";

		public LegacySettingsProfileReadResult Read(LegacySettingsProfileCandidate candidate)
		{
			if (candidate == null)
			{
				throw new ArgumentNullException("candidate");
			}
			try
			{
				XmlDocument xmlDocument = new XmlDocument();
				xmlDocument.XmlResolver = null;
				using (FileStream fileStream = new FileStream(candidate.FilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
				{
					if (fileStream.Length > 262144)
					{
						return LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.TooLarge, "The legacy settings profile exceeds the 256 KiB safety limit.", null);
					}
					using (XmlReader reader = XmlReader.Create(fileStream, CreateSecureReaderSettings()))
					{
						xmlDocument.Load(reader);
					}
				}
				return ParseDocument(candidate, xmlDocument);
			}
			catch (XmlException ex)
			{
				ProjectData.SetProjectError(ex);
				XmlException failureException = ex;
				LegacySettingsProfileReadResult result = LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.MalformedXml, "The legacy settings profile is not safe, well-formed XML.", failureException);
				ProjectData.ClearProjectError();
				return result;
			}
			catch (IOException ex2)
			{
				ProjectData.SetProjectError(ex2);
				IOException failureException2 = ex2;
				LegacySettingsProfileReadResult result = LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.Unavailable, "The legacy settings profile could not be read.", failureException2);
				ProjectData.ClearProjectError();
				return result;
			}
			catch (UnauthorizedAccessException ex3)
			{
				ProjectData.SetProjectError(ex3);
				UnauthorizedAccessException failureException3 = ex3;
				LegacySettingsProfileReadResult result = LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.Unavailable, "Access to the legacy settings profile was denied.", failureException3);
				ProjectData.ClearProjectError();
				return result;
			}
			catch (SecurityException ex4)
			{
				ProjectData.SetProjectError(ex4);
				SecurityException failureException4 = ex4;
				LegacySettingsProfileReadResult result = LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.Unavailable, "The legacy settings profile was blocked by the security policy.", failureException4);
				ProjectData.ClearProjectError();
				return result;
			}
		}

		private static XmlReaderSettings CreateSecureReaderSettings()
		{
			return new XmlReaderSettings
			{
				DtdProcessing = DtdProcessing.Prohibit,
				XmlResolver = null,
				MaxCharactersInDocument = 262144,
				MaxCharactersFromEntities = 0,
				IgnoreComments = true,
				IgnoreProcessingInstructions = true
			};
		}

		private static LegacySettingsProfileReadResult ParseDocument(LegacySettingsProfileCandidate candidate, XmlDocument document)
		{
			if (document.DocumentElement != null && Operators.CompareString(document.DocumentElement.Name, "configuration", false) == 0 && document.DocumentElement.NamespaceURI.Length == 0)
			{
				XmlNodeList xmlNodeList = document.SelectNodes("/configuration/userSettings");
				XmlNodeList xmlNodeList2 = document.SelectNodes("/configuration/userSettings/Compact_Cassette_Catalogue.My.MySettings");
				if (xmlNodeList != null && xmlNodeList.Count == 1 && xmlNodeList2 != null && xmlNodeList2.Count == 1)
				{
					XmlNode xmlNode = xmlNodeList2.Item(0);
					HashSet<string> hashSet = new HashSet<string>(StringComparer.Ordinal);
					string defaultDirectory = null;
					UpdateCheckPolicy updatePolicy = UpdateCheckPolicy.Never;
					DateTime minValue = DateTime.MinValue;
					IEnumerator enumerator = default(IEnumerator);
					bool hasLastUpdateCheck = default(bool);
					bool hasUpdatePolicy = default(bool);
					bool hasDefaultDirectory = default(bool);
					bool showMessages = default(bool);
					bool hasShowMessages = default(bool);
					int num = default(int);
					try
					{
						enumerator = xmlNode.ChildNodes.GetEnumerator();
						while (enumerator.MoveNext())
						{
							XmlNode xmlNode2 = (XmlNode)enumerator.Current;
							if (xmlNode2.NodeType == XmlNodeType.Element && Operators.CompareString(xmlNode2.Name, "setting", false) == 0)
							{
								XmlElement xmlElement = (XmlElement)xmlNode2;
								string attribute = xmlElement.GetAttribute("name");
								if (IsKnownSetting(attribute))
								{
									if (!hashSet.Add(attribute))
									{
										return LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.DuplicateSetting, "The legacy profile contains a duplicate supported setting: " + attribute, null);
									}
									if (Operators.CompareString(xmlElement.GetAttribute("serializeAs"), "String", false) != 0)
									{
										return InvalidValue(candidate, "The legacy setting has an unsupported serialization mode: " + attribute);
									}
									XmlNodeList xmlNodeList3 = xmlElement.SelectNodes("value");
									if (xmlNodeList3 != null && xmlNodeList3.Count == 1 && !HasElementChild(xmlNodeList3.Item(0)))
									{
										string innerText = xmlNodeList3.Item(0).InnerText;
										if (Operators.CompareString(attribute, "showMessages", false) != 0)
										{
											if (Operators.CompareString(attribute, "defaultDirectory", false) != 0)
											{
												if (Operators.CompareString(attribute, "checkUpdates", false) != 0)
												{
													if (Operators.CompareString(attribute, "lastUpdateCheck", false) == 0)
													{
														if (!TryParseLastUpdateCheck(innerText, ref minValue))
														{
															return InvalidValue(candidate, "lastUpdateCheck is not a supported date value.");
														}
														hasLastUpdateCheck = true;
													}
												}
												else
												{
													if (!TryParseUpdatePolicy(innerText, ref updatePolicy))
													{
														return InvalidValue(candidate, "checkUpdates is not a supported policy value.");
													}
													hasUpdatePolicy = true;
												}
											}
											else
											{
												if (innerText.Length > 32768)
												{
													return InvalidValue(candidate, "defaultDirectory exceeds the supported safety limit.");
												}
												defaultDirectory = innerText;
												hasDefaultDirectory = true;
											}
										}
										else
										{
											if (!bool.TryParse(innerText.Trim(), out showMessages))
											{
												return InvalidValue(candidate, "showMessages is not a Boolean value.");
											}
											hasShowMessages = true;
										}
										num = checked(num + 1);
										continue;
									}
									return InvalidValue(candidate, "The legacy setting does not contain one scalar value: " + attribute);
								}
							}
						}
					}
					finally
					{
						if (enumerator is IDisposable)
						{
							(enumerator as IDisposable).Dispose();
						}
					}
					if (num == 0)
					{
						return InvalidStructure(candidate, "The legacy section contains no supported C3 settings.");
					}
					return LegacySettingsProfileReadResult.Succeeded(new LegacyUserSettingsProfile(candidate, hasShowMessages, showMessages, hasDefaultDirectory, defaultDirectory, hasUpdatePolicy, updatePolicy, hasLastUpdateCheck, minValue));
				}
				return InvalidStructure(candidate, "The exact C3 1.x user settings section was not found once.");
			}
			return InvalidStructure(candidate, "The configuration root is not the legacy C3 root.");
		}

		private static bool IsKnownSetting(string name)
		{
			if (Operators.CompareString(name, "showMessages", false) != 0 && Operators.CompareString(name, "defaultDirectory", false) != 0 && Operators.CompareString(name, "checkUpdates", false) != 0)
			{
				return Operators.CompareString(name, "lastUpdateCheck", false) == 0;
			}
			return true;
		}

		private static bool HasElementChild(XmlNode node)
		{
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = node.ChildNodes.GetEnumerator();
				while (enumerator.MoveNext())
				{
					if (((XmlNode)enumerator.Current).NodeType == XmlNodeType.Element)
					{
						return true;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return false;
		}

		private static bool TryParseUpdatePolicy(string rawValue, ref UpdateCheckPolicy updatePolicy)
		{
			if (bool.TryParse((rawValue ?? string.Empty).Trim(), out bool flag))
			{
				updatePolicy = (flag ? UpdateCheckPolicy.Startup : UpdateCheckPolicy.Never);
				return true;
			}
			return UpdateCheckSchedule.TryParseStored(rawValue, ref updatePolicy);
		}

		private static bool TryParseLastUpdateCheck(string rawValue, ref DateTime lastUpdateCheck)
		{
			string text = (rawValue ?? string.Empty).Trim();
			if (text.Length == 0)
			{
				lastUpdateCheck = DateTime.MinValue;
				return true;
			}
			return DateTime.TryParse(text, CultureInfo.InvariantCulture, DateTimeStyles.AllowLeadingWhite | DateTimeStyles.AllowTrailingWhite | DateTimeStyles.AllowInnerWhite | DateTimeStyles.RoundtripKind, out lastUpdateCheck);
		}

		private static LegacySettingsProfileReadResult InvalidStructure(LegacySettingsProfileCandidate candidate, string message)
		{
			return LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.InvalidStructure, message, null);
		}

		private static LegacySettingsProfileReadResult InvalidValue(LegacySettingsProfileCandidate candidate, string message)
		{
			return LegacySettingsProfileReadResult.Failed(candidate, LegacySettingsProfileReadFailure.InvalidValue, message, null);
		}
	}
}
